pkg/app/api/service/webservice/service.proto

```diff
@@ -401,3 +404,13 @@ message ListAPIKeysRequest {
 message ListAPIKeysResponse {
     repeated model.APIKey keys = 1;
 }
+
+message GetApplicationDeploymentFrequencyRequest {
+    string application_id = 1 [(validate.rules).string.min_len = 1];
+    repeated int64 dates = 2;
+}
+
+message GetApplicationDeploymentFrequencyResponse {
+    // key: date, value: deployment_count_of_the_day
+    map<int64, int64> deployment_counts = 1;
+}
```

Comment:
I think we can abstract the models/definitions around this `Insights` functionality.
First of all, a graph can be considered as a series of data points.
So maybe only one `rpc` to retrieve the graph data is enough.

The web client sends a request including:
- metrics_kind: what kind of metrics should be returned (deployment frequency, change failure rate, MTTR...)
- step: the time unit (daily, weekly, monthly, yearly)
- range: from when to when
- filters: e.g. application id...

and the response will contain either a list of data points or an error.

Here are suggestions:

- inside the model package

``` proto
message InsightDataPoint {
   int64 timestamp
   float value
}

enum InsightMetricsKind {
   DEPLOYMENT_FREQUENCY
   CHANGE_FAILURE_RATE
   ...
}

enum InsightStep {
   DAILY
   WEEKLY
   ...
}

```

- inside the service package

``` proto
rpc GetInsightData()...

message GetInsightDataRequest{
   InsightMetricsKind metrics_kind;
   ...
}

message GetInsightDataResponse {
    int64 updated_at;
    repeated InsightDataPoint data_points;
}

```

Reply:
Looks good. Some metrics kind may not be able to be represented by only timestamp and value alone, but then let's reconsider it at that time.

Reply:
Good :)
fixed like above.